wikimedia/wikipedia-ios
# Improve logic for push notification tap routing

**Phabricator:** 
https://phabricator.wikimedia.org/T287628

### Notes
This makes the push notification tap routing a little bit smarter, taking into account the user's editing state and presenting a confirmation alert before routing. It also no longer pops to the main navigation controller's root before routing, so we're no longer wiping out their rabbit hole.

### Test Steps
This is easiest to test by triggering fake notifications to a test device through our internal push utility tool. Reach out to me if you need help setting that up. Once you're able to send notifications to device, use these steps to test various configurations. If you think of any others, feel free to test those too.

1. Open up Settings modal. Trigger push. Tap push. Settings should dismiss, then Notifications Center pushed on from the right.
2. Open up Notifications Center from Explore. Trigger push. Tap push. Nothing should happen (no push), because the user is already on Notifications Center.
3. From History tab, create a deep deep rabbit hole by tapping various article links. Trigger push. Tap Push. Notifications Center should push on from the right, preserving the rabbit hole.
4. Set up a couple of presenting modals by setting up languages on Search (Settings > Search > Show languages on Search), then go to the Search tab > More > Add another language. Trigger push. Tap push. All modals should dismiss, then Notifications Center pushed on from the right. 
5. Set up a smaller modal. Go to the Places tab. You should see one of our smaller modals asking for Location Services. Trigger push. Tap push. Smaller modal should dismiss and Notifications Center pushed on from the right.
6. Now try Editor flow. Tap the 2nd or later Edit pencil on an article. Tap through editor onboarding. While on editor screen, trigger push notification. Tap notification. You should see an alert confirmation. Tapping cancel dismisses confirmation and keeps the user in Editor. Tapping Yes dismisses all modals (including editing) and pushes on Notification Center from the right.
7. Repeat step 6, but do it from farther in editing flow (the edit Preview screen, or the Edit summary screen before the user publishes). You can also try tapping the Insert link keyboard toolbar button or Insert Media keyboard toolbar button and triggering it from the subsequent modals. In all these cases, the same alert view and it's behavior should happen from step 6.
8. Try editing article descriptions - Tap the 1st Edit pencil in an article, then choose "Edit article description". Go past onboarding modal if necessary. Trigger notification, tap notification. You should see the same alert and behavior from step 6 at this point.
